File: frontend/src/api/PostService.js
import axios from "axios";
import {
  setError,
  setLoading,
  setSuccess,
  setPostList,
} from "../redux/postSlice";

const PostService = {
  getPost: async (postId) =>
    axios
      .get("api/v1/posts/" + postId)
      .then((res) => res.data)
      .catch((e) => e.response.data),

  getPosts: async () => axios.get("api/v1/posts").then((res) => res.data),

  reactPost: async ({ postId, type }) =>
    axios
      .post("api/v1/reactions", { postId, type })
      .then((res) => res.data)
      .catch((e) => e.response.data),

  deleteReactPost: async (postId) =>
    axios
      .delete("api/v1/reactions/post/" + postId)
      .then((res) => res.data)
      .catch((e) => e.response.data),

  createPost: async (caption, file, sharedPostId) =>
    axios
      .post(
        "api/v1/posts",
        { caption, file, sharedPostId },
        { headers: { "Content-Type": "multipart/form-data" } }
      )
      .then((res) => res.data)
      .catch((e) => e.response.data),

  editPost: async (postId, caption, file) =>
    axios
      .put(
        "api/v1/posts/" + postId,
        { caption, file },
        { headers: { "Content-Type": "multipart/form-data" } }
      )
      .then((res) => res.data)
      .catch((e) => e.response.data),

  removePost: async (postId) =>
    axios
      .delete("/api/v1/posts/" + postId)
      .then((res) => res.data)
      .catch((e) => e.response.data),

  getPostByUserId: async (userId) =>
    axios.get("api/v1/posts/user/" + userId).then((res) => res.data),
};

export const {
  getPost,
  getPosts,
  reactPost,
  deleteReactPost,
  createPost,
  editPost,
  removePost,
  getPostByUserId,
} = PostService;

File: backend/src/main/java/com/example/socinet/service/PostService.java
package com.example.socinet.service;

import com.example.socinet.dto.PostDto;
import com.example.socinet.entity.Post;
import com.example.socinet.repository.PostRepository;
import com.example.socinet.security.AccountDetail;
import com.example.socinet.util.Helper;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class PostService {
    private final PostRepository postRepo;
    private final FirebaseStorageService storageService;
    private final long MAX_IMAGE_SIZE = 3 * 1024 * 1024; // 3MB
    private final long MAX_VIDEO_SIZE = 15 * 1024 * 1024; //15MB

    public List<PostDto> getPosts(int page, int size){
        List<PostDto> postsDto = new ArrayList<>();
        postRepo.findAll().forEach(post -> postsDto.add(0, new PostDto(post)));
        return postsDto;
    }

    public PostDto getPost(Long id) throws Exception{
        Optional<Post> post = postRepo.findById(id);
        if(post.isEmpty()) throw new Exception("No post found");
        return new PostDto(post.get());
    }

    public List<PostDto> getPostsByUserId(Long userId, Pageable pageable) throws Exception{
        List<Post> posts = postRepo.findByUser_Id(userId, pageable);
        List<PostDto> postsDto = new ArrayList<>();
        posts.forEach(post -> postsDto.add(0, new PostDto(post)));
        return postsDto;
    }

    public PostDto createPost(String caption, MultipartFile image, MultipartFile video, Long sharedPostId) throws Exception{
        AccountDetail accountDetail = Helper.getAccountDetail();
        Post sharedPost = null;
        if(sharedPostId != null) {
            Optional<Post> sharedPostOptional = postRepo.findById(sharedPostId);
            if(sharedPostOptional.isEmpty()) throw new Exception("Shared post not found");
            else sharedPost = sharedPostOptional.get();
        }

        // Upload file lên firebase
        String imageUrl = "";
        String videoUrl = "";
        if(image != null && image.getSize() <= MAX_IMAGE_SIZE){
            imageUrl = storageService.upload("images", image);
        }

        if(video != null && video.getSize() <= MAX_VIDEO_SIZE){
            videoUrl = storageService.upload("videos", video);
        }

        Post post = Post.builder()
                .caption(caption)
                .user(accountDetail.getUser())
                .sharedPost(sharedPost)
                .imageUrl(imageUrl)
                .videoUrl(videoUrl)
                .isActive(true)
                .build();
        return new PostDto(postRepo.save(post));
    }

    public PostDto updatePost(Long postId, String caption, MultipartFile image, MultipartFile video) throws Exception{
        AccountDetail accountDetail = Helper.getAccountDetail();
        Optional<Post> postOptional = postRepo.findByIdAndUser_Id(postId, accountDetail.getUser().getId());
        if(postOptional.isEmpty()) throw new Exception("No post found");
        Post post = postOptional.get();
        post.setCaption(caption);
        if(image != null){
            String imageUrl = storageService.upload("images", image);
            post.setImageUrl(imageUrl);
        }
        if(video != null){
            String videoUrl = storageService.upload("images", video);
            post.setImageUrl(videoUrl);
        }
        return new PostDto(postRepo.save(post));
    }

    public void deleteUserPost(Long postId) throws Exception {
        Long userId = Helper.getUserId();
        Optional<Post> post = postRepo.findByIdAndUser_Id(postId, userId);
        if(post.isEmpty()) throw new Exception("No post found");
        postRepo.delete(post.get());
    }

}
